Error while running flask with eventlet

I'm running flask with eventlet and I have the following error
```
AttributeError: 'Response' object has no attribute 'status_code'
```
In raven I could change transport to `EventletHttpTransport` but I can't see this class in the current SDK

python version: 3.6.1
flask vesrion: 0.12.2
sdk version: 0.7.3
gunicorn version: 19.9.0

```
[2019-02-26 23:37:55 +0200] [80579] [ERROR] Error handling request /socket.io/?EIO=3&transport=websocket
Traceback (most recent call last):
  File "/Users/eric/.pyenv/versions/3.6.1/envs/romee/lib/python3.6/site-packages/gunicorn/workers/base_async.py", line 56, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/Users/eric/.pyenv/versions/3.6.1/envs/romee/lib/python3.6/site-packages/gunicorn/workers/base_async.py", line 116, in handle_request
    resp.close()
  File "/Users/eric/.pyenv/versions/3.6.1/envs/romee/lib/python3.6/site-packages/gunicorn/http/wsgi.py", line 409, in close
    self.send_headers()
  File "/Users/eric/.pyenv/versions/3.6.1/envs/romee/lib/python3.6/site-packages/gunicorn/http/wsgi.py", line 325, in send_headers
    tosend = self.default_headers()
  File "/Users/eric/.pyenv/versions/3.6.1/envs/romee/lib/python3.6/site-packages/gunicorn/http/wsgi.py", line 306, in default_headers
    elif self.should_close():
  File "/Users/eric/.pyenv/versions/3.6.1/envs/romee/lib/python3.6/site-packages/gunicorn/http/wsgi.py", line 229, in should_close
    if self.status_code < 200 or self.status_code in (204, 304):
AttributeError: 'Response' object has no attribute 'status_code' 
```
running gunicron with 1 worker (instead of 4) every thing seems to work fine

killing gunicron procsses raise this exception
```
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/Users/eric/.pyenv/versions/3.6.1/envs/romee/lib/python3.6/site-packages/sentry_sdk/worker.py", line 66, in flush
    self._wait_flush(timeout, callback)
  File "/Users/eric/.pyenv/versions/3.6.1/envs/romee/lib/python3.6/site-packages/sentry_sdk/worker.py", line 71, in _wait_flush
    if not self._timed_queue_join(initial_timeout):
  File "/Users/eric/.pyenv/versions/3.6.1/envs/romee/lib/python3.6/site-packages/sentry_sdk/worker.py", line 33, in _timed_queue_join
    queue.all_tasks_done.acquire()
AttributeError: 'Queue' object has no attribute 'all_tasks_done'
```

--- a/sentry_sdk/worker.py
+++ b/sentry_sdk/worker.py
@@ -52,9 +52,10 @@ class BackgroundWorker(object):
         if real_all_tasks_done is not None:
             real_all_tasks_done.acquire()
             all_tasks_done = real_all_tasks_done  # type: Optional[Any]
-        else:
-            # eventlet
+        elif queue.__module__.startswith("eventlet."):
             all_tasks_done = getattr(queue, "_cond", None)
+        else:
+            all_tasks_done = None
 
         try:
             while queue.unfinished_tasks:  # type: ignore

--- a/sentry_sdk/worker.py
+++ b/sentry_sdk/worker.py
@@ -1,7 +1,7 @@
 import os
 
-import threading
-import time
+from threading import Thread, Lock
+from time import sleep, time
 from sentry_sdk._compat import queue, check_thread_support
 from sentry_sdk.utils import logger
 
@@ -23,8 +23,8 @@ class BackgroundWorker(object):
         # type: () -> None
         check_thread_support()
         self._queue = queue.Queue(-1)  # type: Queue[Any]
-        self._lock = threading.Lock()
-        self._thread = None  # type: Optional[threading.Thread]
+        self._lock = Lock()
+        self._thread = None  # type: Optional[Thread]
         self._thread_for_pid = None  # type: Optional[int]
 
     @property
@@ -43,7 +43,7 @@ class BackgroundWorker(object):
 
     def _timed_queue_join(self, timeout):
         # type: (float) -> bool
-        deadline = time.time() + timeout
+        deadline = time() + timeout
         queue = self._queue
 
         real_all_tasks_done = getattr(
@@ -58,7 +58,7 @@ class BackgroundWorker(object):
 
         try:
             while queue.unfinished_tasks:  # type: ignore
-                delay = deadline - time.time()
+                delay = deadline - time()
                 if delay <= 0:
                     return False
                 if all_tasks_done is not None:
@@ -76,7 +76,7 @@ class BackgroundWorker(object):
         # type: () -> None
         with self._lock:
             if not self.is_alive:
-                self._thread = threading.Thread(
+                self._thread = Thread(
                     target=self._target, name="raven-sentry.BackgroundWorker"
                 )
                 self._thread.setDaemon(True)
@@ -128,4 +128,4 @@ class BackgroundWorker(object):
                     logger.error("Failed processing job", exc_info=True)
             finally:
                 self._queue.task_done()
-            # time.sleep(0)
+            sleep(0)

--- a/sentry_sdk/worker.py
+++ b/sentry_sdk/worker.py
@@ -1,7 +1,7 @@
 import os
 
-from threading import Thread, Lock
-from time import sleep, time
+import threading
+import time
 from sentry_sdk._compat import queue, check_thread_support
 from sentry_sdk.utils import logger
 
@@ -23,8 +23,8 @@ class BackgroundWorker(object):
         # type: () -> None
         check_thread_support()
         self._queue = queue.Queue(-1)  # type: Queue[Any]
-        self._lock = Lock()
-        self._thread = None  # type: Optional[Thread]
+        self._lock = threading.Lock()
+        self._thread = None  # type: Optional[threading.Thread]
         self._thread_for_pid = None  # type: Optional[int]
 
     @property
@@ -43,24 +43,40 @@ class BackgroundWorker(object):
 
     def _timed_queue_join(self, timeout):
         # type: (float) -> bool
-        deadline = time() + timeout
+        deadline = time.time() + timeout
         queue = self._queue
-        queue.all_tasks_done.acquire()  # type: ignore
+
+        real_all_tasks_done = getattr(
+            queue, "all_tasks_done", None
+        )  # type: Optional[Any]
+        if real_all_tasks_done is not None:
+            real_all_tasks_done.acquire()
+            all_tasks_done = real_all_tasks_done  # type: Optional[Any]
+        else:
+            # eventlet
+            all_tasks_done = getattr(queue, "_cond", None)
+
         try:
             while queue.unfinished_tasks:  # type: ignore
-                delay = deadline - time()
+                delay = deadline - time.time()
                 if delay <= 0:
                     return False
-                queue.all_tasks_done.wait(timeout=delay)  # type: ignore
+                if all_tasks_done is not None:
+                    all_tasks_done.wait(timeout=delay)
+                else:
+                    # worst case, we just poll the number of remaining tasks
+                    sleep(0.1)
+
             return True
         finally:
-            queue.all_tasks_done.release()  # type: ignore
+            if real_all_tasks_done is not None:
+                real_all_tasks_done.release()  # type: ignore
 
     def start(self):
         # type: () -> None
         with self._lock:
             if not self.is_alive:
-                self._thread = Thread(
+                self._thread = threading.Thread(
                     target=self._target, name="raven-sentry.BackgroundWorker"
                 )
                 self._thread.setDaemon(True)
@@ -112,4 +128,4 @@ class BackgroundWorker(object):
                     logger.error("Failed processing job", exc_info=True)
             finally:
                 self._queue.task_done()
-            sleep(0)
+            # time.sleep(0)

--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -6,3 +6,4 @@ Werkzeug==0.14.1
 pytest-localserver==0.4.1
 pytest-cov==2.6.0
 gevent
+eventlet

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -4,6 +4,9 @@ import json
 
 import pytest
 
+import gevent
+import eventlet
+
 import sentry_sdk
 from sentry_sdk._compat import reraise, string_types, iteritems
 from sentry_sdk.transport import Transport
@@ -238,3 +241,29 @@ class EventStreamReader(object):
 
     def read_flush(self):
         assert self.file.readline() == b"flush\n"
+
+
+# scope=session ensures that fixture is run earlier
+@pytest.fixture(scope="session", params=[None, "eventlet", "gevent"])
+def maybe_monkeypatched_threading(request):
+    if request.param == "eventlet":
+        try:
+            eventlet.monkey_patch()
+        except AttributeError as e:
+            if "'thread.RLock' object has no attribute" in str(e):
+                # https://bitbucket.org/pypy/pypy/issues/2962/gevent-cannot-patch-rlock-under-pypy-27-7
+                pytest.skip("https://github.com/eventlet/eventlet/issues/546")
+            else:
+                raise
+    elif request.param == "gevent":
+        try:
+            gevent.monkey.patch_all()
+        except Exception as e:
+            if "_RLock__owner" in str(e):
+                pytest.skip("https://github.com/gevent/gevent/issues/1380")
+            else:
+                raise
+    else:
+        assert request.param is None
+
+    return request.param

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -100,6 +100,9 @@ def _capture_internal_warnings():
         if "Something has already installed a non-asyncio" in str(warning.message):
             continue
 
+        if "dns.hash" in str(warning.message) or "dns/namedict" in warning.filename:
+            continue
+
         raise AssertionError(warning)
 
 

--- a/tests/test_transport.py
+++ b/tests/test_transport.py
@@ -4,14 +4,21 @@ import pickle
 
 from datetime import datetime
 
+import eventlet
 import pytest
 
-from sentry_sdk import Hub, Client, add_breadcrumb, capture_message
+
+@pytest.fixture(scope="session", params=[True, False])
+def eventlet_maybe_patched(request):
+    if request.param:
+        eventlet.monkey_patch()
 
 
 @pytest.fixture(params=[True, False])
 def make_client(request):
     def inner(*args, **kwargs):
+        from sentry_sdk import Client
+
         client = Client(*args, **kwargs)
         if request.param:
             client = pickle.loads(pickle.dumps(client))
@@ -22,8 +29,21 @@ def make_client(request):
 
 
 @pytest.mark.parametrize("debug", (True, False))
-def test_transport_works(httpserver, request, capsys, caplog, debug, make_client):
+@pytest.mark.parametrize("client_flush_method", ["close", "flush"])
+def test_transport_works(
+    httpserver,
+    request,
+    capsys,
+    caplog,
+    debug,
+    make_client,
+    client_flush_method,
+    eventlet_maybe_patched,
+):
     httpserver.serve_content("ok", 200)
+
+    from sentry_sdk import Hub, Client, add_breadcrumb, capture_message
+
     caplog.set_level(logging.DEBUG)
 
     client = make_client(
@@ -34,7 +54,8 @@ def test_transport_works(httpserver, request, capsys, caplog, debug, make_client
 
     add_breadcrumb(level="info", message="i like bread", timestamp=datetime.now())
     capture_message("löl")
-    client.close()
+
+    getattr(client, client_flush_method)()
 
     out, err = capsys.readouterr()
     assert not err and not out

--- a/tests/test_transport.py
+++ b/tests/test_transport.py
@@ -10,6 +10,7 @@ import pytest
 from sentry_sdk import Hub, Client, add_breadcrumb, capture_message
 
 
+# scope=session ensures that fixture is run earlier
 @pytest.fixture(scope="session", params=[True, False])
 def eventlet_maybe_patched(request):
     if request.param:

--- a/tests/test_transport.py
+++ b/tests/test_transport.py
@@ -4,22 +4,9 @@ import pickle
 
 from datetime import datetime
 
-import eventlet
 import pytest
 
 from sentry_sdk import Hub, Client, add_breadcrumb, capture_message
-
-
-# scope=session ensures that fixture is run earlier
-@pytest.fixture(scope="session", params=[True, False])
-def eventlet_maybe_patched(request):
-    if request.param:
-        try:
-            eventlet.monkey_patch()
-        except AttributeError as e:
-            if "'thread.RLock' object has no attribute" in str(e):
-                # https://bitbucket.org/pypy/pypy/issues/2962/gevent-cannot-patch-rlock-under-pypy-27-7
-                pytest.skip("https://github.com/eventlet/eventlet/issues/546")
 
 
 @pytest.fixture(params=[True, False])
@@ -44,7 +31,7 @@ def test_transport_works(
     debug,
     make_client,
     client_flush_method,
-    eventlet_maybe_patched,
+    maybe_monkeypatched_threading,
 ):
     httpserver.serve_content("ok", 200)
 

--- a/tests/test_transport.py
+++ b/tests/test_transport.py
@@ -14,7 +14,12 @@ from sentry_sdk import Hub, Client, add_breadcrumb, capture_message
 @pytest.fixture(scope="session", params=[True, False])
 def eventlet_maybe_patched(request):
     if request.param:
-        eventlet.monkey_patch()
+        try:
+            eventlet.monkey_patch()
+        except AttributeError as e:
+            if "'thread.RLock' object has no attribute" in str(e):
+                # https://bitbucket.org/pypy/pypy/issues/2962/gevent-cannot-patch-rlock-under-pypy-27-7
+                pytest.skip("https://github.com/eventlet/eventlet/issues/546")
 
 
 @pytest.fixture(params=[True, False])

--- a/tests/test_transport.py
+++ b/tests/test_transport.py
@@ -7,6 +7,8 @@ from datetime import datetime
 import eventlet
 import pytest
 
+from sentry_sdk import Hub, Client, add_breadcrumb, capture_message
+
 
 @pytest.fixture(scope="session", params=[True, False])
 def eventlet_maybe_patched(request):
@@ -17,8 +19,6 @@ def eventlet_maybe_patched(request):
 @pytest.fixture(params=[True, False])
 def make_client(request):
     def inner(*args, **kwargs):
-        from sentry_sdk import Client
-
         client = Client(*args, **kwargs)
         if request.param:
             client = pickle.loads(pickle.dumps(client))
@@ -41,8 +41,6 @@ def test_transport_works(
     eventlet_maybe_patched,
 ):
     httpserver.serve_content("ok", 200)
-
-    from sentry_sdk import Hub, Client, add_breadcrumb, capture_message
 
     caplog.set_level(logging.DEBUG)
 

--- a/tests/utils/test_contextvars.py
+++ b/tests/utils/test_contextvars.py
@@ -1,37 +1,18 @@
 import random
 import time
 
-import pytest
-import gevent
-
 
 from sentry_sdk.utils import _is_threading_local_monkey_patched
 
 
-def try_gevent_patch_all():
-    try:
-        gevent.monkey.patch_all()
-    except Exception as e:
-        if "_RLock__owner" in str(e):
-            pytest.skip("https://github.com/gevent/gevent/issues/1380")
-        else:
-            raise
+def test_thread_local_is_patched(maybe_monkeypatched_threading):
+    if maybe_monkeypatched_threading is None:
+        assert not _is_threading_local_monkey_patched()
+    else:
+        assert _is_threading_local_monkey_patched()
 
 
-def test_gevent_is_patched():
-    try_gevent_patch_all()
-    assert _is_threading_local_monkey_patched()
-
-
-def test_gevent_is_not_patched():
-    assert not _is_threading_local_monkey_patched()
-
-
-@pytest.mark.parametrize("with_gevent", [True, False])
-def test_leaks(with_gevent):
-    if with_gevent:
-        try_gevent_patch_all()
-
+def test_leaks(maybe_monkeypatched_threading):
     import threading
 
     # Need to explicitly call _get_contextvars because the SDK has already

